lib/querystring.js

```diff
@@ -198,63 +211,179 @@ QueryString.stringify = QueryString.encode = function(obj, sep, eq, options) {
 
   if (typeof qs !== 'string' || qs.length === 0) {
     return obj;
   }
 
+  if (typeof sep !== 'string')
+    sep += '';
+
+  const eqLen = eq.length;
+  const sepLen = sep.length;
+
   var maxKeys = 1000;
   if (options && typeof options.maxKeys === 'number') {
     maxKeys = options.maxKeys;
   }
 
-  // maxKeys <= 0 means that we should not limit keys count
-  if (maxKeys > 0) {
-    qs = qs.split(sep, maxKeys);
-  } else {
-    qs = qs.split(sep);
-  }
-
-  var len = qs.length;
+  var pairs = Infinity;
+  if (maxKeys > 0)
+    pairs = maxKeys;
 
   var decode = QueryString.unescape;
   if (options && typeof options.decodeURIComponent === 'function') {
     decode = options.decodeURIComponent;
   }
-
-  var keys = [];
-  for (var i = 0; i < len; ++i) {
-    // replacePlus() is used instead of a regexp because it is ~15-30% faster
-    // with v8 4.7
-    const x = replacePlus(qs[i]);
-    const idx = x.indexOf(eq);
-    var k, v;
-
-    if (idx >= 0) {
-      k = decodeStr(x.substring(0, idx), decode);
-      v = decodeStr(x.substring(idx + eqLen), decode);
+  const customDecode = (decode !== qsUnescape);
+
+  const keys = [];
+  var lastPos = 0;
+  var sepIdx = 0;
+  var eqIdx = 0;
+  var key = '';
+  var value = '';
+  var keyEncoded = customDecode;
+  var valEncoded = customDecode;
+  var encodeCheck = 0;
+  for (var i = 0; i < qs.length; ++i) {
+    const code = qs.charCodeAt(i);
+
+    // Try matching key/value pair separator (e.g. '&')
+    if (code === sep.charCodeAt(sepIdx)) {
+      if (++sepIdx === sepLen) {
+        // Key/value pair separator match!
+        const end = i - sepIdx + 1;
+        if (eqIdx < eqLen) {
+          // If we didn't find the key/value separator, treat the substring as
+          // part of the key instead of the value
+          if (lastPos < end)
+            key += qs.slice(lastPos, end);
+        } else if (lastPos < end)
+          value += qs.slice(lastPos, end);
+        if (keyEncoded)
+          key = decodeStr(key, decode);
+        if (valEncoded)
+          value = decodeStr(value, decode);
+        // Use a key array lookup instead of using hasOwnProperty(), which is
+        // slower
+        if (keys.indexOf(key) === -1) {
+          obj[key] = value;
+          keys.push(key);
+        } else if (obj[key] instanceof Array) {
+          // `instanceof Array` is used instead of Array.isArray() because it
+          // is ~15-20% faster with v8 4.7 and is safe to use because we are
+          // using it with values being created within this function
+          obj[key].push(value);
+        } else {
+          obj[key] = [obj[key], value];
+        }
+        if (--pairs === 0)
+          break;
+        keyEncoded = valEncoded = customDecode;
+        encodeCheck = 0;
+        key = value = '';
+        lastPos = i + 1;
+        sepIdx = eqIdx = 0;
+      }
+      continue;
     } else {
-      k = decodeStr(x, decode);
-      v = '';
+      sepIdx = 0;
+      if (!valEncoded) {
```

Comment:
Objection withdrawn.  V8's full-codegen emits a COMPARE_IC with strict equality and a TO_BOOLEAN_IC with negation (for this code), making it a wash.  In theory it can emit the equivalent of `bool === isolate->false_value` but it doesn't.